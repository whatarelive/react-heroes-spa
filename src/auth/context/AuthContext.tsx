import { createContext } from "react";
import { State } from '../types/types';

type AuthContextType = {
    state: State,
    login: CallableFunction,
    logout: CallableFunction
}

const intialContext = { 
    state: { 
        logged: false, 
        name: '' 
    }, 
    login: () => {},
    logout: () => {}
}

export const AuthContext = createContext<AuthContextType>( intialContext );