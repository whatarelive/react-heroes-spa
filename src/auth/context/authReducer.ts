import { AuthAction, State, types } from "../types/types";


export const authReducer = ( initialState: State, action: AuthAction ): State => {

    switch ( action?.type ) {
        case types.login:
            return {
                ...initialState,
                logged: true,
                name: action.payload?.name
            };
        
        case types.logout: 
            return {
                logged: false,
            };
    
        default:
            return initialState;
    }
}