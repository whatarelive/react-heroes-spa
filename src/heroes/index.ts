export * from './pages';
export * from './helpers';
export * from './components';
export * from './routes/HeroesRoutes';