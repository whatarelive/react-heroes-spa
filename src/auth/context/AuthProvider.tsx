import { useReducer } from "react"
import { AuthContext } from "./AuthContext";
import { authReducer } from "./authReducer";
import { State } from "../types/types";


const initialstate: State = {
    logged: false,
}

const init = (): State => {
    const savedState = localStorage.getItem('sesion');

    return ( savedState !== null ) ? JSON.parse( savedState ) : initialstate;   
}


export const AuthProvider = ({children}: { children: JSX.Element[] | JSX.Element }) => {

    const [state, dispatch] = useReducer( authReducer, initialstate, init );


  return (
    <AuthContext.Provider value={{ state, dispatch }}>
        { children }
    </AuthContext.Provider>
  )
}
