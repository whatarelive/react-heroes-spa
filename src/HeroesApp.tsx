import { AuthProvider } from "./auth/context/AuthProvider"
import { AppRouter } from "./router/AppRouter"


// Punto de entrada de la aplicaciÃ³n.
export const HeroesApp = () => {

  return (
    <AuthProvider>
      <AppRouter/>
    </AuthProvider>
  )
}
