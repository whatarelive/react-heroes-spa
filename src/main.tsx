import React from 'react'
import ReactDOM from 'react-dom/client'
import { BrowserRouter } from "react-router-dom";
import { HeroesApp } from './HeroesApp'

const root = document.getElementById('root');

if ( root instanceof HTMLDivElement ) {
  
    ReactDOM.createRoot( root ).render(
      <React.StrictMode>
        <BrowserRouter>
          <HeroesApp />
        </BrowserRouter>
      </React.StrictMode>
    )
}
