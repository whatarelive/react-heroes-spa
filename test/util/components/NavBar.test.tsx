import { fireEvent, render, screen } from "@testing-library/react"
import { Navbar } from "../../../src/util"
import { AuthContext } from "../../../src/auth/context/AuthContext"
import { MemoryRouter } from 'react-router'


// Nota: Siempre que trabajamos con un contexto se 
//       debe agregar al renderizar junto con el <MemoryRouter/>.

const mockedUseNavigate = jest.fn();

// Como crear un mock de libreria de react
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate
}));


describe('Pruebas en el componente <NavBar />', () => {

    // Creamos un contexto de ejemplo para evaluar el componente.
    const contextValue = {
        state: {
            logged: true,
            id: 'ABC',
            name: 'Strider'
        },
        login: jest.fn(),
        logout: jest.fn()
    }

    // Para reiniciar las funciones para cada test independiente.
    beforeEach(() => jest.clearAllMocks());

    // Test #1: Sobre la UI completa del componente.
    test('Snapshot del componente cuando se muestra el nombre', () => {
        // Renderizamos el componente.
        const { container } = render( 
            <AuthContext.Provider value={ contextValue }>
                <MemoryRouter>
                    <Navbar/>
                </MemoryRouter>
            </AuthContext.Provider>
        );

        // Esperamos que se realice la comprobacion con el Snapshot anterior.
        expect( container ).toMatchSnapshot();
    })


    // Test #2: Sobre la información que debe mostrar la UI.
    test('debe de mostrar el nombre del usuario', () => {
        // Renderiamos el componente.
        render(
            <AuthContext.Provider value={ contextValue }>
                <MemoryRouter>
                    <Navbar />
                </MemoryRouter>
            </AuthContext.Provider>
        );

        // Esperamos que el nombre del usuario se muestre correctamente.
        expect( screen.getByText( contextValue.state.name ) ).toBeTruthy();
    })


    // Test #3: Sobre el funcionamiento del botón del Logout.
    test('debe de llamar el logout y navigate cuando se hace click en el botón', () => {
        // Renderizamos el componente.
        render(
            <AuthContext.Provider value={ contextValue }>
                <MemoryRouter>
                    <Navbar/>
                </MemoryRouter>
            </AuthContext.Provider>
        );

        // Extraemos el botón del componente renderizado.
        const logoutButton = screen.getByRole('button');
        // Disparamos un evento de click en el botón.
        fireEvent.click( logoutButton );

        // Esperamos que la función de logout asignada al botón sea llamada.
        expect( contextValue.logout ).toHaveBeenCalled();
        // Esperamos que la función navigate sea llamada con los siguientes valores.
        expect( mockedUseNavigate ).toHaveBeenCalledWith('/login', {'replace': true});
    })
})