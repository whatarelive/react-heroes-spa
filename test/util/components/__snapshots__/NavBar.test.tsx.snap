// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <NavBar /> Snapshot del componente cuando se muestra el nombre 1`] = `
<div>
  <nav
    class="navbar navbar-expand-sm navbar-dark bg-dark p-2"
  >
    <div
      class="container-fluid"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        Asociaciones
      </a>
      <div
        class="collapse navbar-collapse"
        id="navbarNavAltMarkup"
      >
        <div
          class="navbar-nav"
        >
          <a
            class="nav-link "
            href="/marvel"
          >
            Marvel
          </a>
          <a
            class="nav-link "
            href="/dc"
          >
            Dc
          </a>
          <a
            class="nav-link "
            href="/search"
          >
            Search
          </a>
        </div>
      </div>
    </div>
    <div
      class="navbar-collapse collapse w-100 order-3 dual-collapse2 d-flex justify-content-end"
    >
      <ul
        class="navbar-nav ml-auto"
      >
        <span
          class="nav-item nav-link text-primary"
        >
          Strider
        </span>
        <button
          class="nav-item nav-link btn"
        >
          Logout
        </button>
      </ul>
    </div>
  </nav>
</div>
`;
